### PR TITLE
wal: do not race reader and writer

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/2423

### CODE REVIEW

File: wal/decoder.go [hunk 1]
```diff
@@ -41,6 +44,9 @@ func newDecoder(rc io.ReadCloser) *decoder {
 }
 
 func (d *decoder) decode(rec *walpb.Record) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
```
Comment (yichengq) [inline wal/decoder.go]: Is this necessary? or for prevention?

**Comment by yichengq** [comment]
Good job to find this out!

**Comment by yichengq** [comment]
lgtm

File: wal/decoder.go (hunk 1 above)
Comment (xiang90) [inline wal/decoder.go]: yea... i am scared. no harm anyway.

**Pull request merged**